### ISSUE TITLE
Feature request: Easyflash save feature

### ISSUE THREAD

**Issue by mambrino** [MiSTer-devel/C64_MiSTer#137]
Easyflash specifications allows to save data on the cartridge (has 1KB SRAM with 256 bytes available and mapped to I/O $DF00-$DFFFF).
This is used more and more in new releases, so, it would be great to support the feature.

These are some games and apps that make use of save functionality, p.e:

https://csdb.dk/release/?id=156948
https://csdb.dk/release/?id=98674

**Comment by sorgelig** [comment]
i can try to add. Just need more technical info. I believe it's not mapped all the time and requires some register to enable/disable

**Comment by sorgelig** [comment]
![image](https://user-images.githubusercontent.com/917089/185742154-fdc7911a-a6fb-491e-851a-820b2571dea5.png)
are you talking about this RAM? It's 256 bytes only. No mention about 1KB paged RAM.

**Comment by mambrino** [comment]
> are you talking about this RAM? It's 256 bytes only. No mention about 1KB paged RAM. There is no mention it's battery backed RAM.

the c64 wiki (https://www.c64-wiki.com/wiki/EasyFlash) says that it has an 8 KiB SRAM installed but only 256 bytes are available.  
It doesn't mention anything about battery backed either, and on the pcb you can't see the battery anywhere, but it must have to keep content when disconnected, no?

**Comment by mambrino** [comment]
> i can try to add. Just need more technical info. I believe it's not mapped all the time and requires some register to enable/disable

Here is the complete repository for EasyFlash with hardware folder including schematics and components bom list (here however says that it's a 64Kbits 6264 SRAM for EasyFlash v1 and 256kbits AS6C62256-55STC SRAM for EasyFlash v3, both cartridges has power pin connected to vdd, nothing about battery backup anywhere):

 https://bitbucket.org/skoe/easyflash/downloads/

**Comment by sorgelig** [comment]
I don't think this RAM is used for saves. It's rather RAM extension.

**Comment by mambrino** [comment]
Easyflash programmer's guide clearly indicates "it can be used to save small portions of code or data", in addition, it's a fact that games like Caren and the Tangled Tentacles, Zak McKracken and Sam's Journey, have an option or key to save the game to cartridge.
There are others like Epyx Game Collection that use this memory to save records, in fact in the main menu there is a tool to dump data from EF to disk and vice versa. (https://csdb.dk/release/?id=107705)
I don't understand what is the point if there is no battery that can keep the data when the computer is turned off, but this could be solved by dumping the content of the crt back to the file system. I think that's what other fpga systems like ultimate II/64 do.

**Comment by paich64** [comment]
as stated on https://skoe.de/easyflash/efintro/   :  
"HOW DOES EASYFLASH WORK?
EasyFlash is an 1 MByte Flash EPROM cartridge with multiple configurations and banks possible. **It does not use traditional UV-EPROMs, but Flash memory**. This makes it possible to “burn” the cartridge using your C64, no EPROM programmer is needed."
and on http://skoe.de/easyflash/recommended/ : 

"**Saving features mentioned here means saving to flash**. If a game has e.g. a high score saving, the game saves the scores directly to the cartridge flash.

A Backup/Restore feature means that the saved data can be exported to disk and imported back to EasyFlash cartridge. This is very useful if you have a cartridge with e.g. saved high scores. You are then able to rewrite the EasyFlash cartridge, and restore the saved data later on."

here is  the programmer guide https://skoe.de/easyflash/files/devdocs/EasyFlash-ProgRef.pdf

**Comment by theenemy-thd** [comment]
+1 for this option .... more and more games are being released on easyflash these days and my group will be releasing more versions of old games converted to run on easyflash with save support for high scores etc. If more info is required on the tech side of saving then please ask and I can ask Mike/knight rider who does the easyflash conversions

**Comment by mambrino** [comment]
[A Pig Quest](https://piggy18.itch.io/a-pig-quest)

Another game released (a big one) in CRT Easyflash format where progress needs to be saved to the cartridge image in order to continue.

There are more and more publications of this type, because due to the size of the game it is not possible to distribute it in floppy disk format.

Please, it would be appreciated if this functionality is added to the core.

**Comment by Schermobianco** [comment]
+1 for this option, thanks guys

**Comment by sorgelig** [comment]
saving is not trivial as it needs emulation of flash memory erase/write cycles.

**Comment by theenemy-thd** [comment]
> saving is not trivial as it needs emulation of flash memory erase/write cycles.

I think how the ultimate 64 gets around it is it allows you in the menu to resave the whole easyflash file in memory, which then includes the highscores etc when reloaded.

Kungfu flash does something different for saving easyflash I think or maybe the same but just overwrites the original file

**Comment by mambrino** [comment]
it could even be an automatic process, just like saving sram on console cores. When the menu opens it could automatically dump the entire easyflash area back into the crt file, wouldn't that be possible?

**Comment by sorgelig** [comment]
it has nothing to do how much of cart will be saved.

**Comment by theenemy-thd** [comment]
So it can't be done then?

**Comment by jedcooper** [comment]
+1 for this option, thanks

**Comment by sorgelig** [comment]
there are 2 issues: 
1) need to emulate flash erase/write commands - should be no problem
2) since there is no dedicated storage for saves, any part of cart can contain saves. Thus entire cart must be saved. With size of 1MB it can take a lot of time...

**Comment by sorgelig** [comment]
SRAM on easyflash is not battery powered as i know. So it cannot hold saves between power cycles.

**Comment by mambrino** [comment]
of course, that's why I tell you to dump that memory area to the crt file, which I suspect is what the original hardware does, since it doesn't have a battery to store between power cycles. Maybe I'm not getting it right, but it seems the only way to do it.

**Comment by sorgelig** [comment]
Original HW has no CPU/MCU on cart and cannot copy anything internally.

**Comment by theenemy-thd** [comment]
When you say "takes a lot of time", I dont think anyone who wants this feature would mind a 5 to 15 second delay on exiting back to menu and it saving. 

couldnt it be done as a on/off option for easyflash save for those who dont care for save .. then it wouldnt effect those individuals.

> there are 2 issues:
> 
> 1. need to emulate flash erase/write commands - should be no problem
> 2. since there is no dedicated storage for saves, any part of cart can contain saves. Thus entire cart must be saved. With size of 1MB it can take a lot of time...

**Comment by sparkysanxion** [comment]
+1 for this feature if it's possible.

it's kinda the only thing holding me back fromn the mister being my main c64 machine at the mo. 
Also +1 for not particularly caring for the save write times too.

Great core tho. brilliant stuff.

**Comment by theenemy-thd** [comment]
Totally agree sanxion, this feature would basically make the mister my goto for c64 ... until then I have to boot my u64 up

**Comment by theenemy-thd** [comment]
Sorry to bang on about this ... but more easyflash games have now been released with saving of progress in game.

Briley witch chronicles 2, space station 23 and yeti mountain

Not having the saving is becoming more of an issue as more easyflash saving games get released.

**Comment by mambrino** [comment]
> Original HW has no CPU/MCU on cart and cannot copy anything internally.

There is no RAM or battery to keep the data, nor does it need a CPU to do so, what the original cartridge does is flashing the eeprom to update the save data area (256 bytes mapped to I/O $DF00-$DFFFF).

as stated by paich64 this howto has info about the process: https://skoe.de/easyflash/efintro/

"HOW DOES EASYFLASH WORK?
EasyFlash is an 1 MByte Flash EPROM cartridge with multiple configurations and banks possible. It does not use traditional UV-EPROMs, but Flash memory. This makes it possible to “burn” the cartridge using your C64, no EPROM programmer is needed."

"Saving features mentioned here means saving to flash. If a game has e.g. a high score saving, the game saves the scores directly to the cartridge flash.

A Backup/Restore feature means that the saved data can be exported to disk and imported back to EasyFlash cartridge. This is very useful if you have a cartridge with e.g. saved high scores. You are then able to rewrite the EasyFlash cartridge, and restore the saved data later on."

**Comment by theenemy-thd** [comment]
The above is not actually correct as the save can be in any 64kb sector of the file hence why a full copy from memory to file would be needed. People don't adhere to keeping the saves to a particular part of the easyflash

**Comment by sparkysanxion** [comment]
Was just going to add (and I know nothing of the technicalities of why/how).... but the way it works on an kung-fu flash on real harware (for those that have not used one) is, when you hit the menu button to go back, if 'something'  has changed, it prompts you to overwrites the full .crt file on the sd card, or not.

**Comment by sorgelig** [comment]
I keep in my mind this feature. So i will implement it when i'll find some good ideas of implementation.

**Comment by xdarkwombatx** [comment]
Yeah this would be huge if implemented.
There are a couple of Ultima remastered games which are on Easyflash but you cant save.

More and more EasyFlash games that require saving are being added.

**Comment by theenemy-thd** [comment]
my usual bump of this subject :D even more easyflash games/versions have been released with saving of some sort, level progress or highs etc ... below the roots, alice in wonderland etc so really wanting this feature more than ever.

**Comment by Scandal-UK** [comment]
For some reason flash memory without a battery confuses some people.

Compact Flash cards have been around for decades, let alone SD cards, SIM cards and even the smart chip on your debit cards. Batteries not included.

**Comment by sparkysanxion** [comment]
Kungfu flash does the same.

**Comment by retro8bits** [comment]
Hello sorgelig,

Excellent core and appreciate all the work that you and all the other contributors have done.

EasyFlash is becoming the de facto standard for cartridges and the c64 community really needs the save functionality.  To move forward, would it be possible to implement the read/write functionality the way that the KungFu Flash 2 does it ? 

https://github.com/KimJorgensen/KungFuFlash2
"CRT files are loaded into RAM, not flash, making large CRT files startup faster. However, changes to EasyFlash cartridges are lost if power is turned off, unless saved via the menu first"

So I would propose in the OSD menu to add two options: 1. Detach cartridge and reset without saving (for those who don't want to save) and 2. Detach cartridge and reset WITH saving.

Kindly let us know.  Thanks and appreciate all the work.

**Comment by retro8bits** [comment]
Hello sorgelig,

I am not technical enough to answer your question.  I was hoping that there would be enough information for you in the KungFu Flash 2 github.  Perhaps someone else in this thread can chime in here and answer your question ?

**Comment by sanstarr** [comment]
Maybe this guide will help?

https://skoe.de/easyflash/files/devdocs/EasyFlash-ProgRef.pdf

Thanks for your support!
Best regards
SanstarR

**Comment by sanstarr** [comment]
Now following terrible, mystical instructions from AI; 

Sorry for that - i am not a software engineer. Only another user, that like to have this easygame save feature.

Thank your very much for your support!


—-


Here’s a concise, PR-ready design summary for MiSTer’s C64 core that implements EasyFlash saving the same way Turbo Chameleon 64 does — i.e., RAM-backed cartridge + EAPI replacement, no real flash emulation.

⸻

EasyFlash (EAPI) Write Support — RAM-backed, TC64-style

Motivation

Many EasyFlash CRTs (e.g., Sam’s Journey) save progress via EAPI. Today, on MiSTer, EF writes change the in-RAM image but persist only after an explicit OSD “save CRT”; that’s expected behavior but EAPI compatibility can be made more robust and predictable by patching in a MiSTer-specific EAPI like Turbo Chameleon does.  ￼

Approach (no real flash emulation)
	•	Do not emulate an AM29F040 command set.
	•	Load the CRT into a mutable RAM buffer, handle EF bank mapping ($DE00) and config ($DE02) per spec.  ￼ ￼
	•	Detect and replace the EAPI binary embedded in the CRT’s first bank with a MiSTer EAPI stub that updates the RAM buffer directly (TC64 does exactly this). Persist modified contents via the core’s “save ROM/CRT” path.  ￼

Spec details you must honor
	•	Registers: $DE00 = bank select (affects ROML/ROMH), $DE02 = control/mapping/LED. (There is no $DE01.)  ￼ ￼
	•	EF-RAM window: $DF00–$DFFF exists (workspace, jump table area used by EAPI). Don’t break visibility.  ￼
	•	Erase granularity: EAPI assumes 64 KiB sectors even if a different chip would offer smaller sectors. Preserve 64 KiB sector semantics in the write path.  ￼
	•	EAPI placement: Many toolchains embed EAPI in hirom0 around $B800; scan bank 0 for the EAPI signature and patch in place.  ￼ ￼

Implementation plan
	1.	Mutable cartridge image

	•	Load EF CRT into a 1 MiB (max) mutable buffer (size from CRT header). Track dirty 64 KiB sectors for efficient persistence.

	2.	Bus/Mapping (unchanged)

	•	Keep current EF decode: map ROML/ROMH with bank from $DE00, apply config bits from $DE02. Make sure EF-RAM $DF00–$DFFF stays readable/writable.  ￼

	3.	EAPI auto-patcher

	•	On EF load, scan bank 0 for the EAPI signature. If found, overwrite the EAPI block with a MiSTer EAPI stub (≈ 700–800 bytes) that implements the standard EAPI entry points (Init, EraseSector, Write, etc.). TC64 replaces EAPI the same way to update its RAM instead of a real flash chip.  ￼
	•	Keep a copy of the original EAPI block in memory (for optional restore on “unload”).

	4.	MiSTer EAPI stub (6502)

	•	Provide a small 6502 library linked into the CRT at patch time that:
	•	Implements EAPIEraseSector(addr) → marks the corresponding 64 KiB region in the cart buffer as erased (set bytes to $FF) and flags it dirty.
	•	Implements EAPIWrite(dst, src, len) → copies len bytes from C64 RAM (src) into the cart buffer at dst (ROM window), enforcing 1→0 bit transitions (no implicit 0→1), and flags dirty.
	•	Returns immediately; no real timing/status toggling required, as the stub is the “driver” (same idea as TC64’s custom EAPI).  ￼
	•	ABI should mirror the public EAPI drivers so existing software keeps working; reference the canonical EAPI repos for symbol names and call patterns.  ￼

	5.	Core ↔ EAPI communication

	•	Because the CPU can’t write to ROM, the stub must request writes via a tiny sideband I/O:
	•	Reserve a private EF-only port (e.g., $DE10…$DE13) while EF mode is active.
	•	Protocol example:
	•	$DE10 = command (0=NOP, 1=ERASE_64K, 2=WRITE_BLOCK)
	•	$DE11/$DE12 = 16-bit ROM offset (within banked window; core resolves to absolute)
	•	$DE13 = length (for small writes; larger chunks in a loop)
	•	The stub writes args, then pokes $DE10 to trigger. Core executes on the mutable cart buffer and auto-clears the command.
	•	This mirrors TC64’s “replace EAPI and talk to the cartridge” concept without exposing global registers; it’s active only under EF.  ￼

	6.	Persistence

	•	Reuse existing OSD item (“Save modified CRT”). On trigger, flush only dirty sectors back to file (or write a full new CRT).
	•	(Optional) Auto-save on core unload behind a setting, default off, to keep current MiSTer expectations. Community guidance already tells users to save the CRT state explicitly.  ￼

Compatibility & limitations
	•	Works with EAPI-based EF CRTs out of the box (that’s most modern EF saves). TC64 does the same: detects EAPI, replaces it, and tells users to save the ROM to persist.  ￼
	•	Raw “flashers” that try to program an actual AM29F040 won’t work (by design). TC64 documents the same limitation.  ￼
	•	Honors EF spec for registers and EF-RAM visibility.  ￼

Test plan
	1.	Sam’s Journey (EF): create a save → reset core (changes remain in RAM); use OSD “Save CRT” → power cycle → confirm save persists.  ￼
	2.	EAPI conformance: run an EAPI test cart (init, erase, write, boundary at 64 KiB).  ￼
	3.	Corner cases: writes crossing bank boundaries; multiple quick saves; mapping flips via $DE02; EF-RAM access unaffected.  ￼

⸻

References
	•	EasyFlash Programming Reference — registers $DE00, $DE02, EF-RAM at $DF00…$DFFF.  ￼
	•	EasySDK/EAPI (drivers & expected ABI).  ￼
	•	Typical EAPI placement (hirom0 @ $B800).  ￼
	•	Turbo Chameleon 64 manual — “EAPI signature found → replace with implementation that updates RAM; save via ROM saver”.  ￼
	•	MiSTer community note — EF saves persist after OSD “save CRT”.  ￼

**Comment by dmantione** [comment]
> do you want to say all games writing to easy flash already support Kung Fu API?

Nope. All games writing to EasyFlash support EAPI. Kung Fu provides games with its own EAPI implementation so games can write to the Kung Fu Flash. If you provide an MiSTer specific EAPI implementation games can write to their "flash memory" using your own proprietari hardware interface.

Going the EAPI way might be less work than implementing a flash ROM in FPGA. However, I do think that recreating a flash ROM is a more future-proof solution than going the EAPI way, because there exist other cartridges with flash memory. If you implement a flash ROM like AM29F040 in FPGA, you could also support writing to Gmod2 cartridges, my Final Cartridge III 101% or my upcoming Super Guppy cartridge.

**Comment by sanstarr** [comment]
Maybe this will help to test?

https://github.com/luigidifraia/easyflash

**Comment by mrdudz** [comment]
https://gitlab.com/skoe/easyflash/-/blob/master/EasySDK/eapi/eapi-am29f040.s?ref_type=heads

this is what you modify to write into your RAM

**Comment by mrdudz** [comment]
> so I don't see why a custom hardware would be much easier.

you already have most of this, you only need a way to make the RAM write able that is used for the emulated cartridge. a couple lines HDL

> but the 1541 Ultimate / Ultimate64 does so too.

Nope. It also uses the EAPI approach - for the reasons mentioned.

> I am not talking about the Gmod2 EEPROM, I am talking about the Gmod2 Flash ROM. While Gmod2 games normally do not write to flash, the Gmod2 absolutely does have a flash ROM accessible from the C64, and by implementing a 29F040 you can use this implementation for Gmod2 as well, achieving full compatibility.

Which is absolutely pointless though - since no game will ever use this (it is not even guaranteed to work on all C64s with real cartridges)

> but you have to add your own EAPI implementation into EasyProg. EasyProg doesn't read the vendor and flash ID, it just tries all EAPIs one-by-one, until EAPInit returns with a succesfull detection.

Ok. Still not useful for testing then :)

**Comment by mrdudz** [comment]
> What I am trying to tell is that if flash location $123456 contains $00 and you write $ff to it, its contents remain $00.

Sure. So in your EAPI erase with $ff and when writing AND the written values. Even that is more trivial in Software than in HDL really.

This is gideons EAPI btw: https://github.com/GideonZ/1541ultimate/blob/master/software/6502/eapi.tas

To me this looks like he really wanted to use this UCI stuff for everything, hence the "29F040" in hardware (its really just doing a few simple things that could just aswell be done in code - it is useless for reuse with any other cartridges for that matter, its tailored for use via this UCI interface)

**Comment by sanstarr** [comment]
Thanks for collecting all this information!

@sorgelig
Maybe this information will help you fulfill this request?

Thanks again!

**Comment by mrdudz** [comment]
> I want to go am29f040 emulation way. Replacing EAPI is easier in software emulator while in FPGA it's more complex.

Having done both before, i have to strongly disagree with this... Unless...well, you don't have a CPU running the menu system for loading cartridges, which would be weird as well :)

**Comment by rsn8887** [comment]
Tested with Sam's Journey retail gmod2-flash file patched to Easyflash via Excess patch - works perfectly! Note: it won't save if you use trainers, that is by design!
(patching tool found here: https://csdb.dk/release/?id=161454)

**Comment by rsn8887** [comment]
Tested with Easyflash converted Sam's Journey - works!

- I used the v1.1 gmod2 to Easyflash patch by Excess from here:
https://csdb.dk/release/?id=172787
- I patched my retail Sam's Journey PAL gmod2-flash.bin file with the .bps patch in the above archive, ignoring checksums, using the online patcher here:
https://www.marcrobledo.com/RomPatcher.js/
(before patching, click on Settings on the bottom and select "fix ROM checksum")
- I saved the resulting file with ending .crt, so I could load it on MiSTer.
- I played on MiSTer until past level 1, selected save in the game, selected save cartridge in the Core menu, reloaded cartridge, and the save game was there
- Note that the Excess EF patch doesn't allow saving when trainers are active, that is by design.
-

**Comment by rsn8887** [comment]
> 1. do saves go into /media/fat/saves/C64 ? (note a C64 directory was not created when I tried to save)

yes, automatically created

> the .zip file contains a file called "Mister".  What do I do with that file or where do I put it on the MisterFPGA ?

You have to either:
- rename your old file `/media/fat/MiSTer` to something else like `/media/fat/MiSTer_old` then ftp the new file onto your MiSTer, then do a cold reboot, otherwise, it will never work. 

or:
 
- Take you sd card out of your MiSTer put it into your PC, and overwrite the MiSTer file with the new one.

> do we keep "Boot EF cartridge" set to YES ?

Doesn't matter according to my experience so far.

**Comment by sorgelig** [comment]
> 3\. do we keep "Boot EF cartridge" set to YES ?

Normally you need to keep it "Yes", otherwise you won't be able to boot it. Use "No" if you need to use external utilities to work with cartridge.

As rsn8887 told already, make sure you replaced MiSTer binary with one from my zip file.

**Comment by sorgelig** [comment]
One important thing to mention: some EF carts don't have full size. Since it's not simply additional bytes but special mapping in memory which is absent for missing cart blocks, MiSTer cannot write to these missing blocks. So make sure cart has full size 1050688 bytes if you plan to write to it.

As mentioned long time ago, since cart is large, the writing time is quite long. I'm not sure if auto-save by opening OSD will be a good option.

**Comment by sanstarr** [comment]
tested with Zak&Maniac Mansion CRT.

**Works perfectly!**

Thank you -VERY- much for your hard work and implementing this feature request!

**Comment by mrdudz** [comment]
> One important thing to mention: some EF carts don't have full size. Since it's not simply additional bytes but special mapping in memory which is absent for missing cart blocks, MiSTer cannot write to these missing blocks. So make sure cart has full size 1050688 bytes if you plan to write to it.

btw, you can "fix" this by converting it to binary and back to crt format using cartconv from VICE:

cartconv -o foo.bin -i foo.crt
cartconv -t ef -b -o fix.crt -i foo.bin

**Comment by mrdudz** [comment]
> you have to replace "ef" with "easy", this is the correct syntax

yes, sorry

**Comment by Scandal-UK** [comment]
> If you don't fix the size, it will crash on game load, but if you fix the size with cartconv, it works.

Yes it is working for me after fixing it this way.

I guess it would be a good idea to ensure all my cart images are 1050688 bytes then? I can just load them from a big zip file anyway..

**Comment by sorgelig** [comment]
Probably i need to auto-expand cart upon loading. Not sure how many trimmed EF carts are floating round.



> A question for Sorgelig: 1. when the core is released will update-all script also update with a new "Mister" binary ? 2. Will the C64 core include updates from the nightly unstable build (i.e C64_unstable_20250815_0575e4.rbf) ?

1. Yes. 2. Of course.

**Comment by sparkysanxion** [comment]
As somone who was hoping for this for a long time....reading all this (as much as I don't particularly understand some of it) has made me happy.
Thanks for all the work put in. Exciting times.

>Probably i need to auto-expand cart upon loading. Not sure how many trimmed EF carts are floating round.

Sounds like a good idea/end user quality of life stuff!

**Comment by Scandal-UK** [comment]
> Probably i need to auto-expand cart upon loading. Not sure how many trimmed EF carts are floating round.

This is the update of the year for me - on the down-side, I can't think of any excuses to buy the C64-Ultimate now.

**Comment by Solo187** [comment]
> Yeah to me it looks like where MultiLoad64 has EasyFlash images they have come from CSDB.

Incorrect. The only crts from CSDB which are in the oneload/multiload64 collection can be found in the EXTRAs folder. These are not made by anyone on the Oneload team.

Everything in the Oneload collection was made by StatMat.
And almost everything in the Multiload64 collection was made from scratch, by myself. With contributions from two other members.

I think the auto expansion of the crts has to be by default either way. As there are many crts like this and you can't expect everybody to create crts in line with one piece of hardwares requirement's. Especially as they all work perfectly with everything else.

**Comment by Scandal-UK** [comment]
> > Yeah to me it looks like where MultiLoad64 has EasyFlash images they have come from CSDB.
> 
> Incorrect. The only crts from CSDB which are in the oneload/multiload64 collection can be found in the EXTRAs folder. These are not made by anyone on the Oneload team.
> 
> Everything in the Oneload collection was made by StatMat. And almost everything in the Multiload collection was made from scratch, by myself. With contributions from two other members.

Ok thanks for the information - yeah the EF games in that collection do have [EasyFlash] in the filename and I have played a lot from the extras folder, so I see what you mean now.

And thanks for all your hard work too - much appreciated.

**Comment by sorgelig** [comment]
new test build:
[c64_efsave1.zip](https://github.com/user-attachments/files/21958133/c64_efsave1.zip)
Support for trimmed EF carts. Saves are now also trimmed.

**Comment by Scandal-UK** [comment]
> new test build: [c64_efsave1.zip](https://github.com/user-attachments/files/21958133/c64_efsave1.zip) Support for trimmed EF carts. Saves are now also trimmed.

WOW. Thank you so much for this!!

**Comment by theenemy-thd** [comment]
I will test new build today

**Comment by sparkysanxion** [comment]
Tested with
Hunters Moon remastered
Briley Witch Chronicles 
International Karate ultimate
Soulforce
Sam’s journey

All working as expected.

Brilliant work!

**Comment by Scandal-UK** [comment]
Yeah all the carts that support cartridge saves on csdb are working - surely time to promote to a nightly build for wider testing?

**Comment by sorgelig** [comment]
I think it's fine for release too. 
I'm thinking if it's a good idea to add an option to save upon OSD opening like it's done for other cores.

**Comment by mambrino** [comment]
> I think it's fine for release too. I'm thinking if it's a good idea to add an option to save upon OSD opening like it's done for other cores.

good idea! and that option can be disabled by default, that way it wouldn't bother anyone.

**Comment by rsn8887** [comment]
I think the issue can be closed now. Since the 08/28 release the core supports this feature.

**Comment by vossstef** [comment]
@mambrino @rsn8887  could it be that the EZFlash CRT save got broken in the meantime ? 

I checked up to date C64 and C128 cores on my Senhor board with the result that 'enabled games' games e.g. BrileyWitchChronicles2-TREX.crt tells 'save failure'.
A check with easyprog-1.8.1.prg reveals that FLASH ID is correctly identified but both erase and program do fail.
Did you encounter similar ?

**Comment by paich64** [comment]
> [@mambrino](https://github.com/mambrino) [@rsn8887](https://github.com/rsn8887) could it be that the EZFlash CRT save got broken in the meantime ?
> 
> I checked up to date C64 and C128 cores on my Senhor board with the result that 'enabled games' games e.g. BrileyWitchChronicles2-TREX.crt tells 'save failure'. A check with easyprog-1.8.1.prg reveals that FLASH ID is correctly identified but both erase and program do fail. Did you encounter similar ?

BrileyWitchChronicles2-TREX.crt is a cracked version which does not even rely on the latest release by the author. Obviously even on one knows what this version has changed. At least you could test / report with a non cracked version (latest 1.03) coming from https://sarahjaneavory.itch.io/briley-witch-chronicles-2-c64

**Comment by paich64** [comment]
> [@mambrino](https://github.com/mambrino) [@rsn8887](https://github.com/rsn8887) could it be that the EZFlash CRT save got broken in the meantime ?
> 
> I checked up to date C64 and C128 cores on my Senhor board with the result that 'enabled games' games e.g. BrileyWitchChronicles2-TREX.crt tells 'save failure'. A check with easyprog-1.8.1.prg reveals that FLASH ID is correctly identified but both erase and program do fail. Did you encounter similar ?

I confirm that 
1) latest https://sarahjaneavory.itch.io/briley-witch-chronicles-2-c64 (1.03) is of easyflash type
2) I confirm saving to cartridge slot fails <==============
3) I confirm also you have the alternative to save to .d64 which works perfectly (loading and saving).

**Comment by vossstef** [comment]
may i assume that the feature was once working and it's now broken ?
I tested also two other in this thread listed games with idential result: save failure

Shall we open an issue ticket to get it fixed ?

**Comment by paich64** [comment]
> may i assume that the feature was once working and it's now broken ? I tested also two other in this thread listed games with idential result: save failure
> 
> Shall we open an issue ticket to get it fixed ?

Maybe test with "known as working" first and if regression confirmed, open a new issue mentionning this one. 

These EF games were reported as "working" : 
Hunters Moon remastered
Briley Witch Chronicles
International Karate ultimate
Soulforce
Sam’s journey

**Comment by paich64** [comment]
> > may i assume that the feature was once working and it's now broken ? I tested also two other in this thread listed games with idential result: save failure
> > Shall we open an issue ticket to get it fixed ?
> 
> Maybe test with "known as working" first and if regression confirmed, open a new issue mentionning this one.
> 
> These EF games were reported as "working" : Hunters Moon remastered Briley Witch Chronicles International Karate ultimate Soulforce Sam’s journey

You would have to identify the exact release this feature was included to confirm it was working. Then identify in which release it started to fail. also keep in mind that you would have to ensure the current main mister is not the culprit also.

**Comment by vossstef** [comment]
@paich64 i got it fixed !

issue was introduced by MD2 support

rtl/cartridge.v

wire [20:0] ezmem_addr = {1'b1, ezrom_addr[19] ? hibanks[ezrom_addr[18:13]] : lobanks[ezrom_addr[18:13]], ezrom_addr[12:0]};
**wire [21:0] ezmem_addr = {2'b10, ezrom_addr[19] ? hibanks[ezrom_addr[18:13]] : lobanks[ezrom_addr[18:13]], ezrom_addr[12:0]};**

to get the new 2Mbyte CRT offset in the DRAM correct

will create a issue and MR
